File: JavaScript/0322.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
 var coinChange = function(coins, amount) {

    // store change and depth
    // the depth is the number of coins
    // initial case is zero change with zero coins
    let coin_change_depth = {0: 0}
    // BFS Queue
    let bfs_coin_queue = [0]

    while (bfs_coin_queue.length) {
        let coin_change = bfs_coin_queue.shift()

        // Reached amount via bfs, guaranteed least nr coins
        if (coin_change === amount) return coin_change_depth[amount]
        
        for (const coin of coins) {
            let next_payment = coin_change + coin
            
            // skip overshoot
            if (next_payment > amount) continue

            // skip payments already tried (part of BFS)
            if (next_payment in coin_change_depth) continue

            // append the dict and queue
            coin_change_depth[next_payment] = coin_change_depth[coin_change] + 1
            bfs_coin_queue.push(next_payment)
        }
    }
    return -1
};

/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
 var coinChange = function(coins, amount) {

    let max = amount+1;
    let dp_cache = new Array(amount+1, max);
    // base case
    dp_cache[0] = 0;
    
    for(let change = 1; change <= amount; change++) 
    {
        let min_coins = max;
        
        for(const coin of coins)
        {
            let prev_change = change - coin;
            
            if(prev_change < 0) continue;
            
            if(dp_cache[prev_change] + 1 < min_coins)
            {
                min_coins = dp_cache[prev_change] +1;
            }            
        }
        dp_cache[change] = min_coins;
    }
    if(dp_cache[amount] === max) return -1;
    return dp_cache[amount];
};

File: CPP/0322.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) 
    {
        ios::sync_with_stdio(false);
        cin.tie(NULL);
        
        // create a cache for the bottom up approach.
        // Zero indexed, and the base case is zero
        // So the length of the cache is amount + 1
        int Max = INT_MAX-1;
        vector<int> dp_cache(amount + 1, Max);
        
        // set the base case
        dp_cache[0] = 0;
        
        // the base case is zero, so we start all next cases at 1
        // till we reach the desired amount
        for(int change=1; change <= amount; change++)
        {
            int min_coins = Max;
            
            for(int coin : coins)
            {
                // going bottom up by trying to find pre calculated values
                int next_change = change - coin;
                
                // skip negative payments
                if (next_change < 0) continue;
                
                // update the minimum count when applicable
                if (dp_cache[next_change] + 1 < min_coins)
                {
                    min_coins = dp_cache[next_change] + 1;
                }
            }
            
            // store the bottom-up result
            dp_cache[change] = min_coins;            
        }
        
        // return the result. if the result is still INT_MAX we did not solve the problem and return -1
        if(dp_cache[amount] == Max) return -1;
        
        return dp_cache[amount];
    }
};

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) 
    {
        ios::sync_with_stdio(false);
        cin.tie(NULL);
        
        // create a cache for the bottom up approach.
        // Zero indexed, and the base case is zero
        // So the length of the cache is amount + 1
        vector<int> dp_cache(amount + 1, INT_MAX-1);
        
        // set the base case
        dp_cache[0] = 0;
        
        // the base case is zero, so we start all next cases at 1
        // till we reach the desired amount
        for(int change=1; change <= amount; change++)
        {           
            for(int coin : coins)
            {
                // going bottom up by trying to find pre calculated values
                int prev_change = change - coin;
                
                // skip negative payments
                if (prev_change < 0) continue;
                
                // update the minimum count when applicable
                if (dp_cache[prev_change] + 1 < dp_cache[change])
                {
                    dp_cache[change] = dp_cache[prev_change] + 1;
                }
            }        
        }
        
        // return the result. if the result is still INT_MAX we did not solve the problem and return -1
        if(dp_cache[amount] == INT_MAX-1) return -1;
        
        return dp_cache[amount];
    }
};
